src/quic/packet.cc

```diff
@@ -0,0 +1,406 @@
+  tracker->TrackField("handle", handle_);
+}
+
+BaseObjectPtr<Packet> Packet::CreateRetryPacket(
+    Environment* env,
+    Listener* listener,
+    const PathDescriptor& path_descriptor,
+    const TokenSecret& token_secret) {
+  auto& random = CID::Factory::random();
+  CID cid = random.Generate();
+  RetryToken token(path_descriptor.version,
+                   path_descriptor.remote_address,
+                   cid,
+                   path_descriptor.dcid,
+                   token_secret);
+  if (!token) return BaseObjectPtr<Packet>();
+
+  const ngtcp2_vec& vec = token;
+
+  size_t pktlen =
+      vec.len + (2 * NGTCP2_MAX_CIDLEN) + path_descriptor.scid.length() + 8;
+
+  auto packet =
+      Create(env, listener, path_descriptor.remote_address, pktlen, "retry");
+  if (!packet) return BaseObjectPtr<Packet>();
+
+  ngtcp2_vec dest = *packet;
+
+  ssize_t nwrite = ngtcp2_crypto_write_retry(dest.base,
+                                             pktlen,
+                                             path_descriptor.version,
+                                             path_descriptor.scid,
+                                             cid,
+                                             path_descriptor.dcid,
+                                             vec.base,
+                                             vec.len);
+  if (nwrite <= 0) return BaseObjectPtr<Packet>();
+  packet->Truncate(nwrite);
+  return packet;
+}
+
+BaseObjectPtr<Packet> Packet::CreateConnectionClosePacket(
+    Environment* env,
+    Listener* listener,
+    const SocketAddress& destination,
+    ngtcp2_conn* conn,
+    const QuicError& error) {
+  auto packet = Packet::Create(
+      env, listener, destination, kDefaultMaxPacketLength, "connection close");
+  ngtcp2_vec vec = *packet;
+
+  ssize_t nwrite = ngtcp2_conn_write_connection_close(
+      conn, nullptr, nullptr, vec.base, vec.len, error, uv_hrtime());
+  if (nwrite < 0) return BaseObjectPtr<Packet>();
+  packet->Truncate(nwrite);
+  return packet;
+}
+
+BaseObjectPtr<Packet> Packet::CreateImmediateConnectionClosePacket(
+    Environment* env,
+    Listener* listener,
+    const SocketAddress& destination,
+    const PathDescriptor& path_descriptor,
+    const QuicError& reason) {
+  auto packet = Packet::Create(env,
+                               listener,
+                               path_descriptor.remote_address,
+                               kDefaultMaxPacketLength,
+                               "immediate connection close (endpoint)");
+  ngtcp2_vec vec = *packet;
+  ssize_t nwrite = ngtcp2_crypto_write_connection_close(
+      vec.base,
+      vec.len,
+      path_descriptor.version,
+      path_descriptor.dcid,
+      path_descriptor.scid,
+      reason.code(),
+      // We do not bother sending a reason string here, even if
+      // there is one in the QuicError
+      nullptr,
+      0);
+  if (nwrite <= 0) return BaseObjectPtr<Packet>();
+  packet->Truncate(static_cast<size_t>(nwrite));
```

Comment:
Is this `static_cast` necessary?

Reply:
> To avoid a compiler warning, yes. `nwrite` is a `ssize_t` and `Truncate` requires `ssize_t`.

Huh? If `nwrite` is already a `ssize_t` and `Truncate` expects a `ssize_t`, then why is a `static_cast` needed?

Reply:
Sorry, that was a typo. `Truncate` expects a `size_t` not `ssize_t`.

Reply:
If `Truncate` expects a `size_t` then why is a `ssize_t` being passed in all (?) other places?

For example:
1. https://github.com/nodejs/node/blob/c9b092f1e98604bd58de08fb4d8de0b9699d4c9c/src/quic/packet.cc#L256-L265
2. https://github.com/nodejs/node/blob/c9b092f1e98604bd58de08fb4d8de0b9699d4c9c/src/quic/packet.cc#L279-L282
3. https://github.com/nodejs/node/blob/c9b092f1e98604bd58de08fb4d8de0b9699d4c9c/src/quic/packet.cc#L340-L346